### ISSUE TITLE
Error running bystro-annotate.pl in Docker

### ISSUE THREAD

**Issue by genesavvy-chris** [akotlar/bystro#85]
I followed the docker instructions in INSTALL.md and I'm trying to annotate a VCF file using bystro-annotate.pl, but I'm hitting this error early into the execution.

> Couldn't open trackNames_meta database. readOnly is set
> [fatal] Couldn't open trackNames_meta database. readOnly is set at /bystro/lib/Seq/DBManager.pm line 1128.

Digging around in other issues, this seems likely to be a metadata database that is probably being used to store metadata about the tracks themselves? I'm having a hard time figuring out whether this failure means I need to review my configuration file (I copied hg38.clean.yml and updated the database_dir and temp_dir per the guidance), or if there is something wrong with my environment.

**Comment by akotlar** [comment]
Ah, what database and Bystro versions are you using?

> This seems likely to be a metadata database that is probably being used to store metadata about the tracks themselves

Yes. The database contains metadata alongside annotation data. This metadata includes a listing of written tracks, the mapping from database indices to tracks, and the mapping from database value indices to track field strings (this allows us to store the entire database as a series of N-d arrays). This error would occur if you had a mismatch between Bystro version/config file and db.

**Comment by genesavvy-chris** [comment]
Where can I find a version number for Bystro? I pulled just a few days ago to start setting this up, so it should be the latest in master. As for the database, I downloaded https://s3.amazonaws.com/bystro-db/hg38.tar.gz.

Here is a more complete description of what I've done so far that may help illuminate the problem.

- Pulled Bystro source
- Copied config/hg38.clean.yml to config/hg38.yml
- Edited config/hg38.yml with paths for database_dir and temp_dir
- Built the docker container
- Pushed container into private registry
- Created a VM running docker and pulled the container image
- Downloaded https://s3.amazonaws.com/bystro-db/hg38.tar.gz and extracted on the VM to the same path I specified in my config
- Added a test VCF file to the VM
- Ran bystro-annotate.pl --config hg38.yml --in <my_vcf_path> --out <my_annotated_vcf_path>

**Comment by akotlar** [comment]
Got it, thanks. Try using the latest release: https://github.com/akotlar/bystro/releases/tag/0.10.2 . Ignore the version number; this is the published version with minor patches; it's a 1.0 except in name.

We're working on cutting a release from master, but there are some changes in database structure that break backward compatibility with the database version you used.

Please let me know if you have any other issues, using this release.
cc @wingolab @Jmeigs1

**Comment by genesavvy-chris** [comment]
Thanks @akotlar -- I was able to get the release running. It turned out there was more going wrong, I'd accidentally extracted my database to a path my container couldn't access. I've resolved that, but I am seeing a different issue now. I double checked my vcf and confirmed it is hg-38 aligned, but I'm getting an error that I think suggests my input may not be lining up with the hg-38 reference.

> Aborted job due to Wrong assembly? chr2: 26735077 not found.

**Comment by akotlar** [comment]
This seems like a perl lib issue.

Do you have this file?
https://github.com/akotlar/bystro/blob/b10/lib/Interface.pm

I would add either run Bystro from within the root Bystro dir (/bystro, so change your WORKDIR to /bystro and then have the CMD as./bin/bystro-annotate.pl), or add /bystro to $PERL5_LIB (export PERL5_LIB=/bystro:$PERL5_LIB, you could add this to environmental variable definitions listed in the docker config).

**Comment by akotlar** [comment]
We will address this detail to make it more clear, cc @wingolab.

**Comment by akotlar** [comment]
I think this also points out that setup should be much easier.
If you keep having issues you could also try  https://bystro.io, which is a hosted version with a backing search engine for filtering, aggregations, and some analysis.

**Comment by akotlar** [comment]
I think I found the issue; the docker build step had an issue in the last release (go module incompatibility with go1.11 and yq). I am testing a fix, will update b10 branch and cut a new release when that's done.

**Comment by akotlar** [comment]
Ok, should be all set. Try 

`docker pull akotlar/bystro:latest`

I'll cut a new release as well. Version will be 0.12.0 to get it on track with remote docker, there are no breaking changes.

**Comment by genesavvy-chris** [comment]
Thanks for the update @akotlar -- I made some more progress thanks to your guidance but I'm unfortunately still not quite reaching success. With regards to the perl libs, I found that because I was deploying my VM in Azure Batch it was overriding the working directory and thus causing the relative paths it was including for dependencies to be incorrect.

I fixed that, and I'm now able to launch successfully but I have a new error:
[fatal] Couldn't find statistics program at bystro-stats at lib/Seq/Statistics.pm line 82.

**Comment by akotlar** [comment]
hm, is ~/go/bin in $PATH?

**Comment by genesavvy-chris** [comment]
PATH=/root/bystro/bin:/usr/local/go/bin:/root/go/bin/:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

**Comment by akotlar** [comment]
Is ~ /root? Can you manually run bystro-stats? That error is generated after failure to find the program using `which`

**Comment by akotlar** [comment]
Using the image I linked above, or alternatively the latest release (here is the repo / commit that this release contains, https://github.com/akotlar/bystro/tree/3d981d8c2e7a51ef3940a1a16a17e181a3b1fb5c) works for me. The previous version had an issue completing go lib installation, because go has been going through a package manager transition.

**Comment by genesavvy-chris** [comment]
Thanks again @akotlar for all the help! Presumably the remaining issues I have are related to pathing being different when I run the container in Batch vs locally, then. I tried running bystro-stats on my container locally and it does find it, so I'll follow up on correcting my paths and I should be good to go :).

**Comment by akotlar** [comment]
Great! Let me know if you have any other issues, we're happy to help!

**Comment by genesavvy-chris** [comment]
I definitely see this as some kind of an environment issue, but I'm quite confused why it continues not to find bystro-stats. To debug I set up a job to run `which bystro-stats` within the context of running the container in Batch, and it does find it: `/root/go/bin//bystro-stats`. I also confirmed home is set to /root.

I noticed the extra slash in the path, but I don't think that should be a problem?

**Comment by akotlar** [comment]
The double slash shouldn't be a problem. This may be specific to Batch. I have a suspicion that it doesn't like the fact that the working directory is /root. What happens if you replace all of the /root mentions with "/"?

Potentially related: https://social.msdn.microsoft.com/Forums/azure/en-US/c42e90b4-25ec-4030-a0ce-24c786da574b/error-running-a-python-script-containerized-file-not-found?forum=azurebatch

**Comment by genesavvy-chris** [comment]
The forum thread you linked was a problem I encountered as well, but I found how to override Batch changing the working directory. I'll try replacing the /root instances with "/" and let you know how it goes. Thanks again!

**Comment by akotlar** [comment]
Sure! Could you also post the docker file and release you're using? Just to be sure, we're on the same page, please make sure you're on the latest release version.

**Comment by genesavvy-chris** [comment]
Okay, so I forked a local branch and synced directly to the commit you linked (https://github.com/akotlar/bystro/tree/3d981d8c2e7a51ef3940a1a16a17e181a3b1fb5c), then I removed all of the /root references from [Dockerfile](https://github.com/akotlar/bystro/files/4967182/Dockerfile.txt). I built a new container image, pushed to my registry and recreated my Batch pool from the new image.

Unfortunately this still hits the same issue finding bystro-stats in Batch. To debug I also set the command line to print out the PATH and to run `which bystro-stats`, but no luck. Is there any way to generate more verbose logs when it fails?

`PATH=/bystro/bin:/usr/local/go/bin:/go/bin/:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`
`/go/bin//bystro-stats`
`[fatal] Couldn't find statistics program at bystro-stats at lib/Seq/Statistics.pm line 82.`

**Comment by akotlar** [comment]
I wonder if `which` isn't working on their platform?

**Comment by akotlar** [comment]
You could try editing that line (82) and removing the which($self->programPath) in favor of $self->programPath

**Comment by akotlar** [comment]
I've attached a modified version that removes the reliance on which, you could edit this file in the docker container and see if it works.

[Statistics-no-which.pm.zip](https://github.com/akotlar/bystro/files/4969100/Statistics-no-which.pm.zip)

**Comment by genesavvy-chris** [comment]
Thanks again @akotlar! I made a small edit to your updated Statistics-no-which.pm file, I noticed when you cut out the `which` the `!` went with it, so I added that back in so we don't fail in the case where it is in fact defined :). I'll give it a shot and let you know what happens!

**Comment by akotlar** [comment]
Sure!  Sorry this has been such a pain, we totally should test against cloud environments besides AWS.

**Comment by genesavvy-chris** [comment]
Still no luck. The one without `which` crashed with this error message:

`Could not load class (Seq::Definition) because : Global symbol "$statsProg" requires explicit package name (did you forget to declare "my $statsProg"?) at lib/Seq/Statistics.pm line 117.
syntax error at lib/Seq/Statistics.pm line 119, near ") {"`

I looked closer at my docker build logs and noticed some of the Perl libraries had also failed to install, so I added a few more of their dependencies into install-perl-libs.sh until I didn't see failures there anymore, but it doesn't seem to have changed anything. With the proper dependencies in place, I went back to the old version of Statistics.pm to try that once more but I got the same error as before (can't find statistics program).

One thing that looks odd to me is I noticed the image you created is quite a bit larger than the ones I am building (1.53 GB for mine, 2.32 GB for yours). I'm not very experienced with Docker to know exactly what to expect here, but the size difference seems suspicious?

**Comment by akotlar** [comment]
The difference between our docker images could be that mine installed all perl libraries and all go libraries. My best guess for why bystro-stats wasn't found was that Go didn't actually install It, hence an image that I knew contained it.

For Statistics.pm a typo in the quick edit I sent you (and you're right it had a missing `!` on 81. Line 115 should be: 
` my $statsProg = $self->programPath;` (notice the trailing semicolon)

**Comment by akotlar** [comment]
Would it be possible to attach the running docker container on Azure and see if you can execute bystro-stats? Else, issue a batch command like `docker bystro:latest bystro-stats`.

Finally if you want, you can try running Bystro without statistics generation (if you're just interested in annotations that is ok), `--run_statistics 0`

**Comment by genesavvy-chris** [comment]
I got some interesting results -- first off not sure if you caught the edit I added to my last post, but my latest round of testing I've gone back to the exact container you provided (the larger one that presumably has the extra libs). Even on that, I was getting the error `[fatal] Couldn't find statistics program at bystro-stats at lib/Seq/Statistics.pm line 82.`. I had thought I was not running statistics by leaving off the `--run_statistics` flag, but I'd missed the default behavior on that one. Adding in `--run_statistics 0` caused it to not crash, but it doesn't seem to have worked, either.

My error log contains the following:
`Use of uninitialized value $firstLine in pattern match (m//) at lib/Seq/Role/IO.pm line 170.`
`Use of uninitialized value $firstLine in pattern match (m//) at lib/Seq/Role/IO.pm line 172.`
`Use of uninitialized value $firstLine in pattern match (m//) at lib/Seq/Role/IO.pm line 174.`
`Use of uninitialized value $header in scalar chomp at lib/Seq.pm line 476.`
`Use of uninitialized value $header in split at lib/Seq.pm line 478.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $_ in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $_ in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $child in string eq at lib/Seq/Headers.pm line 194.`
`Use of uninitialized value $out[0] in join or string at lib/Seq/Headers.pm line 128.`
`Use of uninitialized value $out[1] in join or string at lib/Seq/Headers.pm line 128.`
`Use of uninitialized value $out[2] in join or string at lib/Seq/Headers.pm line 128.`
`Use of uninitialized value in hash element at lib/Seq/Headers.pm line 105.`
`Use of uninitialized value in hash element at lib/Seq/Headers.pm line 105.`
`Use of uninitialized value in hash element at lib/Seq/Headers.pm line 105.`

It did generate an output file, but it appears to be only the column headers and no actual data.

I can connect remotely to the VM that hosts the Azure Batch node and launch the container there, but it's not exactly the same as the way Batch launches the actual tasks so it isn't a perfect mirror of the execution environment. Every time I've tried to run bystro-stats there, it finds the program (but in that context I don't have the files I need available locally to really run it. Taking that investigation one step further, I added `which bystro-stats;` to the beginning of my command line I pass into the container, and the logs showed that even that works in Batch (it finds it in /bin/go//bystro-stats) -- so it doesn't seem to be an inherent problem with `which`, and it doesn't seem to be the case that the file is missing. It only seems to fail to find it from within the context of the process Bystro is running in?
